### ISSUE TITLE
babel-relay-plugin doesn't know about __typename

### ISSUE THREAD

**Issue by devknoll** [facebook/relay#268]
`__typename` is [implicit](https://facebook.github.io/graphql/#sec-Type-Name-Introspection), but babel-relay-plugin doesn't know about it and won't compile fragments containing it (partially related to #162?)

It seems like it might be beneficial to have, otherwise developers will have to manually expose one themselves any time they are composing more than one child fragment:

```js
fragment on Post {
  __typename,
  ${Video.getFragment('video')},
  ${Picture.getFragment('picture')}
}
```

(p.s. would be a lovely time for @dschafer to jump in and drop some news feed knowledge again, since that has since disappeared from the slack archives :wink:)

**Comment by josephsavona** [comment]
Products should *definitely* be able to ask for `__typename` - @dschafer @leebyron thoughts on whether the JSON -> schema builder should generate these fields, or whether we should just hardcode handling for them in the plugin?

**Comment by vladar** [comment]
Also faced with this problem.

Looks like babel plugin code will have to be altered anyway. One of the problems is that `Union` types do not have fields. But they can have this magic `__typename` field.

And plugin code doesn't handle this case.

I just played with it - and simply adding `__typename` to Introspection results doesn't fix the problem. Plugin fails with error `undefined is not a function` in the `printField` function of `GraphQLPrinter` on following line:

```javascript
var fieldDecl = types.getNamedType(type).getFields()[fieldName];
```

(line 347 on my checkout)

**Comment by josephsavona** [comment]
@vladar Yeah, that error indicates that the type of the `__typename` field isn't found - this would also have to be provided by the `graphql` toolchain.

**Comment by devknoll** [comment]
Shall I move this to facebook/graphql-js then, @josephsavona?

**Comment by leebyron** [comment]
__typename is not a field of a type, but a special metafield (hence why its allowed on unions which do not define fields at all) - tools that operate on GraphQL like Relay need to be aware of this metafield.

I think the solution here is for Relay's compiler to be aware of this metafield so that it behaves correctly.

**Comment by josephsavona** [comment]
@leebyron :+1: thanks for confirming

**Comment by devknoll** [comment]
Fixed by #280
